### ISSUE TITLE
TCP Transport without Block Request

### ISSUE THREAD

**Issue by ChrisCuts** [obgm/libcoap#774]
Hello,

I am using libcoap 4.3.0 and I want to use CoAP TCP communication and saw that the messages are separated in blocks like in UDP transport. To enable TCP managed blocks, to increase the transport speed, I tried to set the MTU to UINT_MAX (coap_session_set_mtu(session, UINT_MAX);), which does work up to a 100000 bytes. With 120000 bytes, the transport fails and I could see a strange behaviour in wireshark. The CSM message is corrupted.

Is that a memory allocation problem? Is that the correct way to disable block transport with TCP?

With kind regards
Chris
![grafik](https://user-images.githubusercontent.com/10007694/143035078-85b582f2-13b6-436c-b6cf-b1e85d3e375d.png)

**Comment by mrdeep1** [comment]
Wireshark has onlt relatively recently been upgraded to include undestanding 7.01 is CSM.  However, it is (3.6.0) still broken in the translation of the CSM options (which are different from the standard CoAP options).  For example, CSM Block-wise-transfer option has a value of 4, and Wireshark is treating this as CoAP Option value 4 which is Etag.  CSM Max-Message-Size is option 2, but being reported as unknown as that (2) has not been defined as a CoAP option.

I do not know why your Wireshark is (effectively) reporting the max block size as 0x100000 (Unknown Option: 10 00 00) - I would be expecting this to be 0x800100 based on the standard libcoap definitions.  Your wireshark repots the information slightly different from mine.

![image](https://user-images.githubusercontent.com/35105050/143068359-b0012e54-fae5-4e81-91cc-ae4e2035c35e.png)

libcoap will reject any PDU that is larger than COAP_DEFAULT_MAX_PDU_RX_SIZE which is 0x800100 (8MiB + 256) as a memory resource protection, but CoAP can theorectically support larger sizes.

As the TCP packets have to go over the wire (as standard (1500) or jumbo frames (9000)) the TCP stack may be limiting you here in what you are seeing as write or read sizes.  By having PDUs that are larger than the wire MTU means that the either the TCP stack or the libcoap stack has to re-assemble the data before it can be processed as a PDU which has its overheads.

Yes, you can save on the CoAP PDU non message data overhead by having bigger PDUs, but the tradeoff is memory copying (which is relatively slow) to get data in the right place.

In summary, data writes of up to 8MiB are supported (beyond that code changes are likely to be required), anything larger requires them to be sent as Block-wise transfers [coap_block(3)](https://libcoap.net/doc/reference/develop/man_coap_block.html) which libcoap can handle internally.

**Comment by ChrisCuts** [comment]
Hi,

thank you for your answer. Good to know that the wireshark deserialization does not work for CSM messages correctly.

After some debugging I figured out, that the socket writes only 110048 bytes, what is returned by coap_send_pdu (line 1287 in net.c / coap_send_internal). Then, a queue node is created and put into the delay queue. I think this queue is never called in a client context. I use coap_io_process, here.


Here an overview:

- I start coap client session with:
  - COAP_BLOCK_USE_LIBCOAP | COAP_BLOCK_SINGLE_BODY
  - COAP_PROTO_TCP
  - MTU size UINT_MAX

- Use coap_new_pdu for pdu creation
- I add 110040 bytes of payload with coap_add_data (to get more than 110048 bytes for the whole message)
- And coap_send to send out data

Sorry for asking, but am I doing something wrong?

**Comment by mrdeep1** [comment]
@ChrisCuts You did find an issue with the subsequent partial write of the PDU - which has been fixed in PR #781.  Otherwise, what you are doing looks fine (thanks - for providing how to reproduce the issue). Please confirm that this fixes your actual issue.

The amount written in each write of the same PDU is down to how hte TCP stack works on your O/S.

**Comment by ChrisCuts** [comment]
Thank you for the fix. It seems that it does not solve my problem. The socket gets only the first 110048 bytes and the rest is not transfered:

```
Nov 30 06:32:27.816 DEBG ***0.0.0.0:0 <-> 172.17.0.1:5683 TCP : new outgoing session
Nov 30 06:32:27.817 DEBG ***0.0.0.0:0 <-> 172.17.0.1:5683 TCP : session ack_timeout set to 2.000
Nov 30 06:32:27.817 DEBG ***0.0.0.0:0 <-> 172.17.0.1:5683 TCP : session max_retransmit set to 2
Nov 30 06:32:27.819 DEBG ***EVENT: 0x1001
Nov 30 06:32:27.819 DEBG ***172.17.0.2:42188 <-> 172.17.0.1:5683 TCP : sending CSM
Nov 30 06:32:27.819 DEBG *  172.17.0.2:42188 <-> 172.17.0.1:5683 TCP : sent 7 bytes
v:1 t:CON c:CSM i:0000 {} [ Max-Message-Size:8388864, Block-wise-Transfer: ]
Nov 30 06:32:27.820 DEBG *  172.17.0.2:42188 <-> 172.17.0.1:5683 TCP : received 7 bytes
v:1 t:CON c:CSM i:0000 {} [ Max-Message-Size:1048576, Block-wise-Transfer: ]
Nov 30 06:32:27.820 DEBG ***172.17.0.2:42188 <-> 172.17.0.1:5683 TCP : session connected
Nov 30 06:32:27.820 DEBG ***EVENT: 0x2001
Nov 30 06:32:27.821 DEBG sending CoAP request:
Nov 30 06:32:27.821 DEBG ** 172.17.0.2:42188 <-> 172.17.0.1:5683 TCP : lg_crcv 0x55cf1800b1d0 initialized
Nov 30 06:32:27.821 DEBG *  172.17.0.2:42188 <-> 172.17.0.1:5683 TCP : sent 110048 bytes
v:1 t:CON c:POST i:1bfc {930ad799} [ Uri-Path:other, Uri-Path:block, Content-Format:application/json ] :: binary data length 110040
Nov 30 06:32:27.853 DEBG ** 172.17.0.2:42188 <-> 172.17.0.1:5683 TCP : mid=0x1bfc: delayed
```

After that, nothing happens. I have a breakpoint in coap_session_write and that is never reached after that log. Should the session manage to put more data into the socket after some time, or is the session informed, that it could push more data into it by the system?

**Comment by mrdeep1** [comment]
I assume that you have made the changes to both files, so now

> Nov 30 06:32:27.821 DEBG *  172.17.0.2:42188 <-> 172.17.0.1:5683 TCP : sent 110048 bytes

is reporting what the TCP stack actually accepted.  With the following change to examples/coap-client.c
````
diff --git a/examples/coap-client.c b/examples/coap-client.c
index 890909b..e3bb435 100644
--- a/examples/coap-client.c
+++ b/examples/coap-client.c
@@ -1426,6 +1426,7 @@ open_session(
     /* Non-encrypted session */
     session = coap_new_client_session(ctx, bind_addr, dst, proto);
   }
+coap_session_set_mtu(session, 0xffffffff);
   return session;
 }
 
````
I get (/tmp/bf can be different lengths using `dd`, and my intial TCP stack capabilities is 163840 bytes)
````
# examples/coap-client -m put -f /tmp/bf coap+tcp://127.0.0.1/example_data -v7
Nov 30 09:08:45.707 DEBG ***0.0.0.0:0 <-> 127.0.0.1:5683 TCP : new outgoing session
Nov 30 09:08:45.707 DEBG sending CoAP request:
Nov 30 09:08:45.707 DEBG ** 0.0.0.0:0 <-> 127.0.0.1:5683 TCP : lg_crcv 0x2447250 initialized
Nov 30 09:08:45.708 DEBG ** 0.0.0.0:0 <-> 127.0.0.1:5683 TCP : mid=0x6248: delayed
Nov 30 09:08:45.708 DEBG timeout is set to 90 seconds
Nov 30 09:08:45.708 INFO ****** Next wakeup time 1271310.319000000
Nov 30 09:08:45.708 DEBG ***EVENT: 0x1001
Nov 30 09:08:45.708 DEBG ***127.0.0.1:38831 <-> 127.0.0.1:5683 TCP : sending CSM
Nov 30 09:08:45.708 DEBG *  127.0.0.1:38831 <-> 127.0.0.1:5683 TCP : sent 7 bytes
v:1 t:CON c:CSM i:0000 {} [ Max-Message-Size:8388864, Block-wise-Transfer: ]
Nov 30 09:08:45.708 DEBG *  127.0.0.1:38831 <-> 127.0.0.1:5683 TCP : received 7 bytes
v:1 t:CON c:CSM i:0000 {} [ Max-Message-Size:8388864, Block-wise-Transfer: ]
Nov 30 09:08:45.708 DEBG ***127.0.0.1:38831 <-> 127.0.0.1:5683 TCP : session connected
Nov 30 09:08:45.708 DEBG ***EVENT: 0x2001
Nov 30 09:08:45.708 DEBG ** 127.0.0.1:38831 <-> 127.0.0.1:5683 TCP : mid=0x6248: transmitted after delay
Nov 30 09:08:45.708 DEBG *  127.0.0.1:38831 <-> 127.0.0.1:5683 TCP : sent 163840 bytes
v:1 t:CON c:PUT i:6248 {01} [ Uri-Path:example_data ] :: binary data length 163832
<<0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000<<. . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 
Nov 30 09:08:45.781 DEBG ** 127.0.0.1:38831 <-> 127.0.0.1:5683 TCP : mid=0x6248: transmitted after delay
Nov 30 09:08:45.781 DEBG *  127.0.0.1:38831 <-> 127.0.0.1:5683 TCP : sent 13 bytes
Nov 30 09:08:45.781 INFO ****** Next wakeup time 1271310.319000000
Nov 30 09:08:45.781 INFO ****** Next wakeup time 1271310.319000000
Nov 30 09:08:45.854 DEBG *  127.0.0.1:38831 <-> 127.0.0.1:5683 TCP : received 3 bytes
v:1 t:CON c:2.04 i:0000 {01} [ ]
Nov 30 09:08:45.855 DEBG ** process incoming 2.04 response:
Nov 30 09:08:45.855 INFO ****** Next wakeup time 0.247000000
Nov 30 09:08:45.855 DEBG ** 127.0.0.1:38831 <-> 127.0.0.1:5683 TCP : lg_crcv 0x2447250 released
Nov 30 09:08:45.855 DEBG ***127.0.0.1:38831 <-> 127.0.0.1:5683 TCP : session closed

````
I do notice a difference between the 2 logs is when `DEBG sending CoAP request:` is reported, so it looks likely to be something down to the way that your client is setting up the session / doing coap_send().

**Comment by mrdeep1** [comment]
I assume that your client keeps calling coap_io_process() until all has been transferred?

**Comment by ChrisCuts** [comment]
Hi,

good idea to use the CLI client. I could not reconstruct my issue and have the same results like you. I will investigate regarding the differences in my code and the client.

Thanks for your help
Chris

**Comment by ChrisCuts** [comment]
I did some further investigations and figured out, that a short delay after session creation leads to my fault. Please see the attached code to reconstruct the issue. Please have a look into the main() at the end of the file.

[no_transport_after_short_delay.zip](https://github.com/obgm/libcoap/files/7632170/no_transport_after_short_delay.zip)

**Comment by mrdeep1** [comment]
The current libcoap port is not multi-thread safe (that work is on the TODO list), so the calling of coap_io_process() unfortunately cannot be done in a separate thread as there is not protected access to the context, session or i/o activity.

I would suggest something along the lines of
````
$ diff -Nau /tmp/test.cpp test.cpp
--- /tmp/test.cpp       2021-12-01 08:27:09.000000000 +0000
+++ test.cpp    2021-12-01 14:37:28.995129480 +0000
@@ -161,7 +160,7 @@
     default:;
     }

-    request_handler = std::thread([] () { request_handler_callback(ctx); });
+//    request_handler = std::thread([] () { request_handler_callback(ctx); });

     return session;
 }
@@ -196,8 +195,16 @@

     send_message(session, send_pdu, &optlist, payload);

+    int timeout = 2 * 1000;
+    while (timeout > 0) {
+           int ret = coap_io_process(ctx, timeout);
+           if (ret < 0)
+                   break;
+           timeout -= ret;
+    }
+
     // Wait for answer
-    std::this_thread::sleep_for(std::chrono::seconds(2));
+    //std::this_thread::sleep_for(std::chrono::seconds(2));

     // Break down
     request_handler.detach();
````

**Comment by ChrisCuts** [comment]
Ah, ok. There we go. Many thanks.

Do you have a tip how to implement it in such a way, that notifications could be received in parallel? My first idea was to add a queue that is processed in the main thread together with the io processing. In this design, the new messages would be delayed, until the socket timeout is exceeded. Do you know, if there is any way to interrupt this externally?

```
main thread:
while 1:
  isRequestInQueue?
    coap_send()
  otherwise:
    coap_io_process()
```

My second approach would be to have a session for the observer, using a queue as mentioned above. 500ms delay for registration would be fine for me. The other transport would be processes in a separated session allowing direct processing.

The timing is quite relevant for my application, since we want to use libcoap to communicate with a data broker in a small server.

**Comment by mrdeep1** [comment]
Have you looked at [coap_io(3)](https://libcoap.net/doc/reference/develop/man_coap_io.html) - in particular method 2 for using coap_io_process()?
This does assume that you have EPOLL available.  You can then se a select() call on the libcoap fd as well as another fd which you can use for communication.  If EPOLL is not available, then you can use coap_io_process_with_fds().

So the process can be "interrupted" by activity on the non-libcoap fd(s).

For simplicity, I would be inclined to have the same session for both Observer and direct communications, but there is no reason for not having two distinct CoAP sessions.

**Comment by mrdeep1** [comment]
There is a function `coap_can_exit(ctx)` which returns `1` if there is nothing outstanding, else `0` if something to do.  It is used in a more complex loop than you are using in examples/coap-client.c while checking if there is any i/o pending.

It is not documented in the man pages, but can be found in the Doxygen ducumentation under https://libcoap.net/doc/reference/develop/group__context.html

**Comment by ChrisCuts** [comment]
Hello back,

thanks for the support. I implemented epoll and this worked out pretty well. I stepped into another problem with transmissions of more than 1 MB over TCP. I set the MTU size to 2MB. During communication, I get the log message 'pdu too big'. It seems, that the client creates a new session in the background after the CSM handshake.

When I add some additional debug logs in coap_pdu_init and coap_pdu_resize, I got the following output:
```
WARN coap_pdu_init
WARN Max size       2097146
WARN Size           2097146

WARN coap_pdu_init
WARN Max size       20
WARN Size           20

WARN coap_pdu_init
WARN Max size       2097146
WARN Size           2097146

WARN coap_pdu_init
WARN Max size       1048570
WARN Size           1048570

WARN coap_pdu_resize
WARN Max size 1048570
WARN New size 1050635
WARN coap_pdu_resize: pdu too big
```

So, I received a message with 1050635 bytes. But the limit is set to 1048570, which stems from the received CSM message (in handle_signaling in net.c). In the RFC the max message size is figured out to define the maximum size the sender (of the CSM) can receive, not the maximum it will send out.
From my perspective, this is not correct implemented. [RFC](https://tools.ietf.org/id/draft-ietf-core-coap-tcp-tls-11.html#max-message-size)

**Comment by mrdeep1** [comment]
On reciept of a CSM Max-Message-Size, the mtu size (mtu is not really the right word) is updated and will then limits any future transmission sizes - following [RFC 8323 Section 5.3.1](https://datatracker.ietf.org/doc/html/rfc8323#section-5.3.1).

However, when the Xmt and Rcv pdu Max-Message-Sizes are different things do not work properly as you have found.

In the current code, it is not always clear whether a PDU is going to be used for Xmt or Rcv when it is created, so trying to differentiate Xmt and Rcv mtu sizes could be tricky in a code update - but this needs to be clarified.

The alternative could be to send another CSM Max-Message-Size if after CSM exchange it is determined that there is a lower common Max-Message-Size that should be used to remove asymmetric differences.  But we may run into implementation support issues here for the subsequent CSM message.

**Comment by ChrisCuts** [comment]
As a work-around for the current implementation, I set COAP_DEFAULT_MAX_PDU_RX_SIZE to the server value. Since we are operating in a closed environment, this works out for me.

**Comment by mrdeep1** [comment]
Setting COAP_DEFAULT_MAX_PDU_RX_SIZE  should be fine in your specific case.

I will need to check into whether BERT is working or not.  From what you have found, it looks like there is an issue.

**Comment by ChrisCuts** [comment]
This is the behaviour of an aiocoap client:
![grafik](https://user-images.githubusercontent.com/10007694/155133755-2cf3eb62-29b8-44e9-82a9-35b48f54759e.png)

This is the behaviour of the libcoap client having COAP_BLOCK_USE_LIBCOAP enabled:
![grafik](https://user-images.githubusercontent.com/10007694/155134044-043a305a-648d-42b2-aa0d-11b2a9bee505.png)

I think the code is in block.c coap_handle_response_send_block(*)
```
        block.num++;
        coap_update_option(pdu, p->option,
                           coap_encode_var_safe(buf, sizeof(buf),
                             (block.num << 4) |
                             ((p->offset + chunk < p->length) << 3) |
                             block.szx),
                           buf);
```

**Comment by mrdeep1** [comment]
See #833 for a fix to the asymmetric Xmt and Rcv CSM Max-Message-Size.

BERT will be handled separately.

**Comment by mrdeep1** [comment]
BERT is not being handled - the main issue is that BERT blocks (comprising of 0 or more chunks of 1024 with possibly the remaining data of 1-1024 bytes_ that fit into the Max-Message-Size are not getting sent as a single item.  The data is sent using 1024 byte blocks at a time.

The easiest way to fix this I think is to have a new coap_block2_t that knows about BERT and have new block2_ functions to minimize the changes to the code.

**Comment by mrdeep1** [comment]
See PR #848 raised to provide BERT support.  Can you please check this out when you get a chance.

**Comment by ChrisCuts** [comment]
Great, thank you!
I will check it in my test environment the next days.

**Comment by mrdeep1** [comment]
Closed as PR #848 has now been merged.
